diff --git a/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java b/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java
index 058678d67d..41c9a1bf3d 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java
@@ -36,6 +36,7 @@ import org.apache.ambari.server.state.svccomphost.ServiceComponentHostOpFailedEv
 import org.apache.ambari.server.state.svccomphost.ServiceComponentHostOpInProgressEvent;
 import org.apache.ambari.server.state.svccomphost.ServiceComponentHostOpSucceededEvent;
 import org.apache.ambari.server.utils.StageUtils;
+import org.apache.ambari.server.utils.VersionUtils;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
@@ -384,6 +385,21 @@ public class HeartBeatHandler {
     String hostname = register.getHostname();
     long now = System.currentTimeMillis();
 
+    String agentVersion = register.getAgentVersion();
+    String serverVersion = ambariMetaInfo.getServerVersion();
+    if (!VersionUtils.areVersionsCompatible(serverVersion, agentVersion)) {
+      LOG.warn("Received registration request from host with non compatible"
+              + " agent version"
+              + ", hostname=" + hostname
+              + ", agentVersion=" + agentVersion
+              + ", serverVersion=" + serverVersion);
+      throw new AmbariException("Cannot register host with non compatible"
+              + " agent version"
+              + ", hostname=" + hostname
+              + ", agentVersion=" + agentVersion
+              + ", serverVersion=" + serverVersion);
+    }
+
     String agentOsType = getOsType(register.getHardwareProfile().getOS(),
         register.getHardwareProfile().getOSRelease());
     if (!ambariMetaInfo.areOsTypesCompatible(
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/api/services/AmbariMetaInfo.java b/ambari-server/src/main/java/org/apache/ambari/server/api/services/AmbariMetaInfo.java
index cab351edc3..49e948332b 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/api/services/AmbariMetaInfo.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/api/services/AmbariMetaInfo.java
@@ -25,7 +25,9 @@ import org.apache.ambari.server.AmbariException;
 import org.apache.ambari.server.ObjectNotFoundException;
 import org.apache.ambari.server.StackAccessException;
 import org.apache.ambari.server.configuration.Configuration;
+import org.apache.ambari.server.resources.ResourceManager;
 import org.apache.ambari.server.state.*;
+import org.apache.ambari.server.state.Stack;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.w3c.dom.*;
@@ -37,12 +39,7 @@ import javax.xml.parsers.ParserConfigurationException;
 import java.io.File;
 import java.io.FilenameFilter;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 
 /**
  * ServiceInfo responsible getting information about cluster.
@@ -50,8 +47,10 @@ import java.util.Set;
 @Singleton
 public class AmbariMetaInfo {
 
+    private String serverVersion = "undefined";
     private List<StackInfo> stacksResult = new ArrayList<StackInfo>();
     private File stackRoot;
+    private File serverVersionFile;
     private final static Logger LOG = LoggerFactory
             .getLogger(AmbariMetaInfo.class);
     
@@ -104,13 +103,16 @@ public class AmbariMetaInfo {
      */
     @Inject
     public AmbariMetaInfo(Configuration conf) throws Exception {
-        String stackPath = conf.getMetadataPath();
-        this.stackRoot = new File(stackPath);
+      String stackPath = conf.getMetadataPath();
+      String serverVersionFilePath = conf.getServerVersionFilePath();
+      this.stackRoot = new File(stackPath);
+      this.serverVersionFile = new File(serverVersionFilePath);
     }
 
     @Inject
-    public AmbariMetaInfo(File stackRoot) throws Exception {
-        this.stackRoot = stackRoot;
+    public AmbariMetaInfo(File stackRoot, File serverVersionFile) throws Exception {
+      this.stackRoot = stackRoot;
+      this.serverVersionFile = serverVersionFile;
     }
 
 
@@ -121,6 +123,7 @@ public class AmbariMetaInfo {
      */
     @Inject
     public void init() throws Exception {
+        readServerVersion();
         getConfigurationInformation(stackRoot);
     }
 
@@ -587,8 +590,16 @@ public class AmbariMetaInfo {
       return resultOperatingSystem;
   }
 
-   
-    
+
+    private void readServerVersion() throws  Exception{
+      File versionFile = this.serverVersionFile;
+      String t = (new File(".")).getAbsolutePath();
+      if (! versionFile.exists()) {
+        throw new AmbariException("Server version file does not exist");
+      }
+      serverVersion = new Scanner(versionFile).useDelimiter("\\Z").next();
+    }
+
     private void getConfigurationInformation(File stackRoot) throws Exception {
 
         if (LOG.isDebugEnabled()) {
@@ -676,6 +687,10 @@ public class AmbariMetaInfo {
         }
     }
 
+  public String getServerVersion() {
+    return serverVersion;
+  }
+
   private StackInfo getStackInfo(File stackVersionFolder) {
 
     StackInfo stackInfo = new StackInfo();
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
index aabd93232a..42b1f82a71 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
@@ -67,6 +67,8 @@ public class Configuration {
   public static final String PASSPHRASE_KEY = "security.server.passphrase";
   public static final String RESOURCES_DIR_KEY = "resources.dir";
   public static final String METADETA_DIR_PATH = "metadata.path";
+  public static final String SERVER_VERSION_FILE = "server.version.file";
+  public static final String SERVER_VERSION_KEY = "version";
 
 
   public static final String CLIENT_SECURITY_KEY = "client.security";
@@ -302,7 +304,7 @@ public class Configuration {
 
   public String getBootSetupAgentScript() {
     return properties.getProperty(BOOTSTRAP_SETUP_AGENT_SCRIPT,
-        "/usr/lib/python2.7/site-packages/ambari_server/setupAgent.py");
+        "/usr/lib/python2.6/site-packages/ambari_server/setupAgent.py");
   }
 
   public String getBootSetupAgentPassword() {
@@ -358,6 +360,10 @@ public class Configuration {
     return properties.getProperty(METADETA_DIR_PATH);
   }
 
+  public String getServerVersionFilePath() {
+    return properties.getProperty(SERVER_VERSION_FILE);
+  }
+
   /**
    * Check to see if the API should be authenticated or not
    * @return false if not, true if the authentication is enabled.
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariServer.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariServer.java
index b66ca88eed..9aa33aa4f4 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariServer.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariServer.java
@@ -36,6 +36,7 @@ import org.apache.ambari.server.bootstrap.BootStrapImpl;
 import org.apache.ambari.server.configuration.Configuration;
 import org.apache.ambari.server.orm.GuiceJpaInitializer;
 import org.apache.ambari.server.orm.PersistenceType;
+import org.apache.ambari.server.orm.dao.MetainfoDAO;
 import org.apache.ambari.server.resources.ResourceManager;
 import org.apache.ambari.server.resources.api.rest.GetResource;
 import org.apache.ambari.server.security.CertificateManager;
@@ -96,6 +97,8 @@ public class AmbariServer {
   Injector injector;
   @Inject
   AmbariMetaInfo ambariMetaInfo;
+  @Inject
+  MetainfoDAO metainfoDAO;
 
   public String getServerOsType() {
     return configs.getServerOsType();
@@ -114,6 +117,8 @@ public class AmbariServer {
     server = new Server();
     serverForAgent = new Server();
 
+    checkDBVersion();
+
     try {
       ClassPathXmlApplicationContext parentSpringAppContext =
           new ClassPathXmlApplicationContext();
@@ -371,6 +376,21 @@ public class AmbariServer {
     }
   }
 
+  protected void checkDBVersion() throws AmbariException {
+    LOG.info("Checking DB store version");
+    String databaseVersion = metainfoDAO.findByKey(Configuration.SERVER_VERSION_KEY).getMetainfoValue();
+    String serverVersion = ambariMetaInfo.getServerVersion();
+    if (! databaseVersion.equals(serverVersion)) {
+      String error = "Current database store version is not compatible with " +
+              "current server version"
+              + ", serverVersion=" + serverVersion
+              + ", databaseVersion=" + databaseVersion;
+      LOG.warn(error);
+      throw new AmbariException(error);
+    }
+    LOG.info("DB store version is compatible");
+  }
+
   public void stop() throws Exception {
     try {
       server.stop();
