diff --git a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
index a427334cef..0e1be7a405 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
@@ -327,7 +327,7 @@ public class Configuration {
 
   public String getBootSetupAgentScript() {
     return properties.getProperty(BOOTSTRAP_SETUP_AGENT_SCRIPT,
-        "/usr/lib/python2.7/site-packages/ambari_server/setupAgent.py");
+        "/usr/lib/python2.6/site-packages/ambari_server/setupAgent.py");
   }
 
   public String getBootSetupAgentPassword() {
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java
index 86eb99399d..cc1ce0d5d6 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java
@@ -125,26 +125,33 @@ public abstract class AbstractControllerResourceProvider extends AbstractResourc
   }
 
   /**
-   * Extracting given query_paramater value from the predicate(parsed http body)
+   * Extracting given query_paramater value from the predicate
    * @param queryParameterId
    * @param predicate
    * @return
    */
   protected static Object getQueryParameterValue(String queryParameterId, Predicate predicate) {
 
+    Object result = null;
+
     if (predicate instanceof ArrayPredicate) {
       ArrayPredicate arrayPredicate  = (ArrayPredicate) predicate;
       for (Predicate predicateItem : arrayPredicate.getPredicates()) {
-        if (predicateItem instanceof ComparisonPredicate) {
-          EqualsPredicate equalsPredicate =
-              (EqualsPredicate) predicateItem;
-          if (queryParameterId.equals(equalsPredicate.getPropertyId())) {
-            return equalsPredicate.getValue();
+          if (predicateItem instanceof EqualsPredicate) {
+            EqualsPredicate equalsPredicate =
+                (EqualsPredicate) predicateItem;
+            if (queryParameterId.equals(equalsPredicate.getPropertyId())) {
+              return equalsPredicate.getValue();
+            }
+          } else {
+            result = getQueryParameterValue(queryParameterId, predicateItem);
+            if (result != null) {
+              return result;
           }
         }
       }
 
     }
-    return null;
+    return result;
   }
 }
