diff --git a/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java b/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java
index df9735047d..861c7f7978 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/agent/HeartBeatHandler.java
@@ -236,17 +236,20 @@ public class HeartBeatHandler {
               State prevState = scHost.getState();
               State liveState = State.valueOf(State.class, status.getStatus());
               if (prevState.equals(State.INSTALLED)
-                      || prevState.equals(State.START_FAILED)
-                      || prevState.equals(State.STARTED)
-                      || prevState.equals(State.STOP_FAILED)) {
+                  || prevState.equals(State.START_FAILED)
+                  || prevState.equals(State.STARTED)
+                  || prevState.equals(State.STARTING)
+                  || prevState.equals(State.STOPPING)
+                  || prevState.equals(State.STOP_FAILED)) {
                 scHost.setState(liveState);
-                if (!prevState.equals(liveState) && scHost.getServiceComponentName().equals(Role.HBASE_MASTER.toString())) {
+                LOG.info("State of service component " + componentName
+                    + " of service " + status.getServiceName()
+                    + " of cluster " + status.getClusterName()
+                    + " has changed from " + prevState + " to " + liveState
+                    + " at host " + hostname);
+                if (!prevState.equals(liveState)
+                    && scHost.getServiceComponentName().equals(Role.HBASE_MASTER.toString())) {
                     scaner.updateHBaseMaster(scHost);
-                  LOG.info("State of service component " + componentName
-                          + " of service " + status.getServiceName()
-                          + " of cluster " + status.getClusterName()
-                          + " has changed from " + prevState + " to " + liveState
-                          + " at host " + hostname);
                 }
               }
 
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
index 38fab71b74..fa77589d8b 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
@@ -282,7 +282,7 @@ public class Configuration {
 
   public String getBootSetupAgentScript() {
     return properties.getProperty(BOOTSTRAP_SETUP_AGENT_SCRIPT,
-        "/usr/lib/python2.7/site-packages/ambari_server/setupAgent.py");
+        "/usr/lib/python2.6/site-packages/ambari_server/setupAgent.py");
   }
 
   public String getBootSetupAgentPassword() {
