diff --git a/ambari-server/src/main/java/org/apache/ambari/server/ldap/service/ads/DefaultLdapConfigurationService.java b/ambari-server/src/main/java/org/apache/ambari/server/ldap/service/ads/DefaultLdapConfigurationService.java
index cffb19027e..47e758cb22 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/ldap/service/ads/DefaultLdapConfigurationService.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/ldap/service/ads/DefaultLdapConfigurationService.java
@@ -181,7 +181,7 @@ public class DefaultLdapConfigurationService implements LdapConfigurationService
     EntryMapper<String> entryMapper = new EntryMapper<String>() {
       @Override
       public String map(Entry entry) throws LdapException {
-        return entry.get(ambariLdapConfiguration.groupNameAttribute()).get().getString();
+        return entry.get(ambariLdapConfiguration.groupNameAttribute()).get().getValue();
       }
     };
 
