diff --git a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
index 38fab71b74..fa77589d8b 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/configuration/Configuration.java
@@ -282,7 +282,7 @@ public class Configuration {
 
   public String getBootSetupAgentScript() {
     return properties.getProperty(BOOTSTRAP_SETUP_AGENT_SCRIPT,
-        "/usr/lib/python2.7/site-packages/ambari_server/setupAgent.py");
+        "/usr/lib/python2.6/site-packages/ambari_server/setupAgent.py");
   }
 
   public String getBootSetupAgentPassword() {
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementController.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementController.java
index 3d10e240ea..4b51d56f78 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementController.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementController.java
@@ -256,24 +256,27 @@ public interface AmbariManagementController {
    *
    * @throws AmbariException thrown if the resource cannot be updated
    */
-  public RequestStatusResponse updateServices(Set<ServiceRequest> requests, Map<String, String> requestProperties)
-      throws AmbariException;
+  public RequestStatusResponse updateServices(Set<ServiceRequest> requests,
+      Map<String, String> requestProperties, boolean runSmokeTest) throws AmbariException;
 
   /**
    * Update the component identified by the given request object with the
    * values carried by the given request object.
    *
    *
+   *
    * @param requests    the request object which defines which component to
    *                   update and the values to set
    *
    * @param requestProperties
+   * @param runSmokeTest
    * @return a track action response
    *
    * @throws AmbariException thrown if the resource cannot be updated
    */
   public RequestStatusResponse updateComponents(
-      Set<ServiceComponentRequest> requests, Map<String, String> requestProperties) throws AmbariException;
+      Set<ServiceComponentRequest> requests, Map<String, String> requestProperties,
+      boolean runSmokeTest) throws AmbariException;
 
   /**
    * Update the host identified by the given request object with the
@@ -292,16 +295,18 @@ public interface AmbariManagementController {
    * values carried by the given request object.
    *
    *
+   *
    * @param requests    the request object which defines which host component to
    *                   update and the values to set
    *
    * @param requestProperties
+   * @param runSmokeTest
    * @return a track action response
    *
    * @throws AmbariException thrown if the resource cannot be updated
    */
   public RequestStatusResponse updateHostComponents(
-      Set<ServiceComponentHostRequest> requests, Map<String, String> requestProperties) throws AmbariException;
+      Set<ServiceComponentHostRequest> requests, Map<String, String> requestProperties, boolean runSmokeTest) throws AmbariException;
   
   /**
    * Updates the users specified.
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java
index 1abcb1da2d..4c520141bf 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/AmbariManagementControllerImpl.java
@@ -1533,7 +1533,7 @@ public class AmbariManagementControllerImpl implements
 
       LOG.info("Creating stages for upgrade.");
       List<Stage> stages = doStageCreation(cluster, changedServices, changedComps, changedScHosts,
-          requestParameters, requestProperties.get(REQUEST_CONTEXT_PROPERTY));
+          requestParameters, requestProperties.get(REQUEST_CONTEXT_PROPERTY), false);
 
       if (stages == null || stages.isEmpty()) {
         return null;
@@ -1716,7 +1716,7 @@ public class AmbariManagementControllerImpl implements
       Map<State, List<Service>> changedServices,
       Map<State, List<ServiceComponent>> changedComps,
       Map<String, Map<State, List<ServiceComponentHost>>> changedScHosts,
-      Map<String, String> requestParameters, String requestContext)
+      Map<String, String> requestParameters, String requestContext, boolean runSmokeTest)
           throws AmbariException {
 
     // TODO handle different transitions?
@@ -1746,7 +1746,7 @@ public class AmbariManagementControllerImpl implements
           continue;
         }
         for (Service s : entry.getValue()) {
-          if (State.INSTALLED == s.getDesiredState()) {
+          if (runSmokeTest && (State.INSTALLED == s.getDesiredState())) {
             smokeTestServices.add(s.getName());
           }
         }
@@ -1789,7 +1789,7 @@ public class AmbariManagementControllerImpl implements
         changedComponentCount.entrySet()) {
       String serviceName = entry.getKey();
       // smoke test service if more than one component is started
-      if (entry.getValue().size() > 1) {
+      if (runSmokeTest && (entry.getValue().size() > 1)) {
         smokeTestServices.add(serviceName);
         continue;
       }
@@ -1801,7 +1801,7 @@ public class AmbariManagementControllerImpl implements
         ComponentInfo compInfo = ambariMetaInfo.getComponentCategory(
             stackId.getStackName(), stackId.getStackVersion(), serviceName,
             componentName);
-        if (compInfo.isMaster()) {
+        if (runSmokeTest && compInfo.isMaster()) {
           smokeTestServices.add(serviceName);
         }
 
@@ -2217,8 +2217,8 @@ public class AmbariManagementControllerImpl implements
 
   @Override
   public synchronized RequestStatusResponse updateServices(
-      Set<ServiceRequest> requests, Map<String, String> requestProperties)
-      throws AmbariException {
+      Set<ServiceRequest> requests, Map<String, String> requestProperties,
+      boolean runSmokeTest) throws AmbariException {
 
     if (requests.isEmpty()) {
       LOG.warn("Received an empty requests set");
@@ -2493,8 +2493,9 @@ public class AmbariManagementControllerImpl implements
 
     Cluster cluster = clusters.getCluster(clusterNames.iterator().next());
 
-    List<Stage> stages = doStageCreation(cluster, changedServices,
-        changedComps, changedScHosts, null, requestProperties.get(REQUEST_CONTEXT_PROPERTY));
+    List<Stage> stages = doStageCreation(cluster, changedServices, changedComps,
+        changedScHosts, null, requestProperties.get(REQUEST_CONTEXT_PROPERTY),
+        runSmokeTest);
     persistStages(stages);
     updateServiceStates(changedServices, changedComps, changedScHosts);
     if (stages == null || stages.isEmpty()) {
@@ -2506,7 +2507,7 @@ public class AmbariManagementControllerImpl implements
 
   @Override
   public synchronized RequestStatusResponse updateComponents(Set<ServiceComponentRequest> requests,
-                                                             Map<String, String> requestProperties)
+                                                             Map<String, String> requestProperties, boolean runSmokeTest)
                                                              throws AmbariException {
 
     if (requests.isEmpty()) {
@@ -2773,7 +2774,7 @@ public class AmbariManagementControllerImpl implements
     Cluster cluster = clusters.getCluster(clusterNames.iterator().next());
 
     List<Stage> stages = doStageCreation(cluster, null,
-        changedComps, changedScHosts, null, requestProperties.get(REQUEST_CONTEXT_PROPERTY));
+        changedComps, changedScHosts, null, requestProperties.get(REQUEST_CONTEXT_PROPERTY), runSmokeTest);
     persistStages(stages);
     updateServiceStates(null, changedComps, changedScHosts);
     if (stages == null || stages.isEmpty()) {
@@ -2855,7 +2856,7 @@ public class AmbariManagementControllerImpl implements
 
   @Override
   public synchronized RequestStatusResponse updateHostComponents(Set<ServiceComponentHostRequest> requests,
-                                                                 Map<String, String> requestProperties)
+                                                                 Map<String, String> requestProperties, boolean runSmokeTest)
                                                                  throws AmbariException {
 
     if (requests.isEmpty()) {
@@ -3147,7 +3148,7 @@ public class AmbariManagementControllerImpl implements
       requestParameters.put(Configuration.UPGRADE_FROM_STACK, gson.toJson(fromStackVersion));
     }
     List<Stage> stages = doStageCreation(cluster, null, null, changedScHosts, requestParameters,
-        requestProperties.get(REQUEST_CONTEXT_PROPERTY));
+        requestProperties.get(REQUEST_CONTEXT_PROPERTY), runSmokeTest);
     persistStages(stages);
     updateServiceStates(null, null, changedScHosts);
     if (stages == null || stages.isEmpty()) {
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java
index 97732df158..86eb99399d 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/AbstractControllerResourceProvider.java
@@ -19,6 +19,10 @@
 package org.apache.ambari.server.controller.internal;
 
 import org.apache.ambari.server.controller.AmbariManagementController;
+import org.apache.ambari.server.controller.predicate.ArrayPredicate;
+import org.apache.ambari.server.controller.predicate.ComparisonPredicate;
+import org.apache.ambari.server.controller.predicate.EqualsPredicate;
+import org.apache.ambari.server.controller.spi.Predicate;
 import org.apache.ambari.server.controller.spi.Resource;
 import org.apache.ambari.server.controller.spi.ResourceProvider;
 
@@ -119,4 +123,28 @@ public abstract class AbstractControllerResourceProvider extends AbstractResourc
         throw new IllegalArgumentException("Unknown type " + type);
     }
   }
+
+  /**
+   * Extracting given query_paramater value from the predicate(parsed http body)
+   * @param queryParameterId
+   * @param predicate
+   * @return
+   */
+  protected static Object getQueryParameterValue(String queryParameterId, Predicate predicate) {
+
+    if (predicate instanceof ArrayPredicate) {
+      ArrayPredicate arrayPredicate  = (ArrayPredicate) predicate;
+      for (Predicate predicateItem : arrayPredicate.getPredicates()) {
+        if (predicateItem instanceof ComparisonPredicate) {
+          EqualsPredicate equalsPredicate =
+              (EqualsPredicate) predicateItem;
+          if (queryParameterId.equals(equalsPredicate.getPropertyId())) {
+            return equalsPredicate.getValue();
+          }
+        }
+      }
+
+    }
+    return null;
+  }
 }
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ComponentResourceProvider.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ComponentResourceProvider.java
index 921b0089c1..e88d8c360f 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ComponentResourceProvider.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ComponentResourceProvider.java
@@ -46,6 +46,9 @@ class ComponentResourceProvider extends AbstractControllerResourceProvider {
   protected static final String COMPONENT_STATE_PROPERTY_ID           = PropertyHelper.getPropertyId("ServiceComponentInfo", "state");
   protected static final String COMPONENT_DESIRED_CONFIGS_PROPERTY_ID = PropertyHelper.getPropertyId("ServiceComponentInfo", "desired_configs");
 
+  //Parameters from the predicate
+  private static final String QUERY_PARAMETERS_RUN_SMOKE_TEST_ID =
+      "params/run_smoke_test";
 
   private static Set<String> pkPropertyIds =
       new HashSet<String>(Arrays.asList(new String[]{
@@ -151,10 +154,13 @@ class ComponentResourceProvider extends AbstractControllerResourceProvider {
       requests.add(compRequest);
     }
 
+    final boolean runSmokeTest = "true".equals(getQueryParameterValue(
+        QUERY_PARAMETERS_RUN_SMOKE_TEST_ID, predicate)) ? true : false;
+
     RequestStatusResponse response = modifyResources(new Command<RequestStatusResponse>() {
       @Override
       public RequestStatusResponse invoke() throws AmbariException {
-        return getManagementController().updateComponents(requests, request.getRequestInfoProperties());
+        return getManagementController().updateComponents(requests, request.getRequestInfoProperties(), runSmokeTest);
       }
     });
 
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/HostComponentResourceProvider.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/HostComponentResourceProvider.java
index 415e4593f3..9412f2c7b6 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/HostComponentResourceProvider.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/HostComponentResourceProvider.java
@@ -63,6 +63,10 @@ class HostComponentResourceProvider extends AbstractControllerResourceProvider {
   protected static final String HOST_COMPONENT_DESIRED_STACK_ID_PROPERTY_ID
       = PropertyHelper.getPropertyId("HostRoles", "desired_stack_id");
 
+  //Parameters from the predicate
+  private static final String QUERY_PARAMETERS_RUN_SMOKE_TEST_ID =
+      "params/run_smoke_test";
+
   private static Set<String> pkPropertyIds =
       new HashSet<String>(Arrays.asList(new String[]{
           HOST_COMPONENT_CLUSTER_NAME_PROPERTY_ID,
@@ -165,6 +169,9 @@ class HostComponentResourceProvider extends AbstractControllerResourceProvider {
     final Set<ServiceComponentHostRequest> requests = new HashSet<ServiceComponentHostRequest>();
     RequestStatusResponse response = null;
 
+    final boolean runSmokeTest = "true".equals(getQueryParameterValue(
+        QUERY_PARAMETERS_RUN_SMOKE_TEST_ID, predicate)) ? true : false;
+
     Iterator<Map<String,Object>> iterator = request.getProperties().iterator();
     if (iterator.hasNext()) {
       for (Map<String, Object> propertyMap : getPropertyMaps(request.getProperties().iterator().next(), predicate)) {
@@ -173,7 +180,7 @@ class HostComponentResourceProvider extends AbstractControllerResourceProvider {
       response = modifyResources(new Command<RequestStatusResponse>() {
         @Override
         public RequestStatusResponse invoke() throws AmbariException {
-          return getManagementController().updateHostComponents(requests, request.getRequestInfoProperties());
+          return getManagementController().updateHostComponents(requests, request.getRequestInfoProperties(), runSmokeTest);
         }
       });
 
diff --git a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ServiceResourceProvider.java b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ServiceResourceProvider.java
index ab488a4f4d..af2e8636d3 100644
--- a/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ServiceResourceProvider.java
+++ b/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ServiceResourceProvider.java
@@ -45,6 +45,9 @@ class ServiceResourceProvider extends AbstractControllerResourceProvider {
   protected static final String SERVICE_SERVICE_STATE_PROPERTY_ID   = PropertyHelper.getPropertyId("ServiceInfo", "state");
   protected static final String SERVICE_DESIRED_CONFIGS_PROPERTY_ID = PropertyHelper.getPropertyId("ServiceInfo", "desired_configs");
 
+  //Parameters from the predicate
+  private static final String QUERY_PARAMETERS_RUN_SMOKE_TEST_ID =
+      "params/run_smoke_test";
 
   private static Set<String> pkPropertyIds =
       new HashSet<String>(Arrays.asList(new String[]{
@@ -138,10 +141,14 @@ class ServiceResourceProvider extends AbstractControllerResourceProvider {
       for (Map<String, Object> propertyMap : getPropertyMaps(iterator.next(), predicate)) {
         requests.add(getRequest(propertyMap));
       }
+
+      final boolean runSmokeTest = "true".equals(getQueryParameterValue(
+          QUERY_PARAMETERS_RUN_SMOKE_TEST_ID, predicate)) ? true : false;
+
       response = modifyResources(new Command<RequestStatusResponse>() {
         @Override
         public RequestStatusResponse invoke() throws AmbariException {
-          return getManagementController().updateServices(requests, request.getRequestInfoProperties());
+          return getManagementController().updateServices(requests, request.getRequestInfoProperties(), runSmokeTest);
         }
       });
     }
